arch/xtensa/core/xtensa-asm2.c

```diff
@@ -226,18 +223,38 @@ void *xtensa_excint1_c(int *interrupted_stack)
 		 * else it will just loop forever
 		 */
 		bsa[BSA_PC_OFF/4] += 3;
-
 	} else {
 		uint32_t ps = bsa[BSA_PS_OFF/4];
+		void *pc = (void *)bsa[BSA_PC_OFF/4];
 
 		__asm__ volatile("rsr.excvaddr %0" : "=r"(vaddr));
 
+		/* Default for exception */
+		int reason = K_ERR_CPU_EXCEPTION;
+
+		/* We need to distinguish between an ill in xtensa_arch_except,
+		 * e.g for k_panic, and any other ill. For exceptions caused by
+		 * xtensa_arch_except calls, we also need to pass the reason_p
+		 * to z_xtensa_fatal_error. Since the ARCH_EXCEPT frame is in the
+		 * BSA, the first arg reason_p is stored at the A2 offset.
+		 * We assign EXCCAUSE the unused, reserved code 63; this may be
+		 * problematic if the app or new boards also decide to repurpose
+		 * this code.
+		 */
+		void *xtensa_arch_except_ill = (void *) (((char *) xtensa_arch_except) + 3);
+
+		if ((pc == xtensa_arch_except_ill) && (cause == 0)) {
+			cause = 63;
+			__asm__ volatile("wsr.exccause %0" : : "r"(cause));
+			reason = bsa[BSA_A2_OFF/4];
```

Comment:
Stuffing this in EXCCAUSE is clever.  I would have guessed this is a read-only register, but the spec says it's writable.  It's sort of a layering bug with our pre-existing dump code, though: we should be extracting the full exception state in one place and then passing it around as regular data instead of relying on special registers.